### ISSUE TITLE
Bug: Generic benchmark class fails for DisassemblyDiagnoser with `Sequence contains no matching element`

### ISSUE THREAD

**Issue by nietras** [dotnet/BenchmarkDotNet#631]
This is I think a known limitation of the current disassembler, but I would really like this to work to avoid having to write a lot of code for different types that I would like to disassemble the code for.

This again is from https://github.com/DotNetCross/Sorting/tree/fc60378eee0cb9cafed9f35d596b42aaaca8269e but the code in question is disabled/uncommented.

I have a base disassembler benchmark class.
```csharp
    [DisassemblyDiagnoser(printAsm: true, printSource: true, recursiveDepth: 4)]
    [SimpleJob(RunStrategy.Monitoring, launchCount: 1, warmupCount: 1, targetCount: 3)]
    //[Config(typeof(SortDisassemblerBenchConfig))]
    public class SortDisassemblerBench<TKey>
    {
        readonly int _length;
        readonly TKey[] _filled;
        readonly TKey[] _work;

        public SortDisassemblerBench(int length, Func<int, TKey> toValue)
        {
            _length = length;
            _filled = new TKey[_length];
            // We use median of three to ensure heap sort code is hit
            new MedianOfThreeKillerSpanFiller().Fill(_filled, _length, toValue);
            _work = new TKey[_length];
        }

        [IterationSetup]
        public void IterationSetup()
        {
            Console.WriteLine($"// {nameof(IterationSetup)} Copy filled to work {_length}");
            Array.Copy(_filled, _work, _length);
        }

        [Benchmark(Baseline = true)]
        public void ArraySort()
        {
            Array.Sort(_work);
        }

        [Benchmark]
        public void SpanSort()
        {
            new Span<TKey>(_work).Sort();
        }
    }
```
That I then have to make a specific type of (something I would also like to avoid):
```csharp
    public class Int32SortDisassemblerBenchNotWorking : SortDisassemblerBench<int>
    {
        public Int32SortDisassemblerBenchNotWorking()
            : base(length: 1024 * 1024, i => i)
        { }
    }
```
And then try to run:
```csharp
BenchmarkRunner.Run<Int32SortDisassemblerBenchNotWorking>();
```
This then fails with:
```
Failed to disassemble with following exception:
Sequence contains no matching element
   at System.Linq.Enumerable.Single[TSource](IEnumerable`1 source, Func`2 predicate)
   at BenchmarkDotNet.Disassembler.Program.Disassemble(Settings settings, ClrRuntime runtime, State state)
   at BenchmarkDotNet.Disassembler.Program.Handle(Settings settings)
   at BenchmarkDotNet.Disassembler.Program.Main(String[] args)
```

**Comment by AndreyAkinshin** [comment]
@adamsitnik, could you please take a look?

**Comment by adamsitnik** [comment]
@nietras Your idea with the workaround has inspired me :)

You suggested a workaround with deriving non-generic type from generic type. I realized that we do  that in our Engine anyway. And I used this to overcome ClrMD limitations

After [this](https://ci.appveyor.com/project/dotnetfoundation/benchmarkdotnet/build/0.10.12.420) build is finished the new package on our CI will contain the fix.

**Comment by nietras** [comment]
@adamsitnik it's good to be a code muse ;-) Thanks for the fix! Will the
build include fix for params ordering too?


On Feb 4, 2018 15:49, "Adam Sitnik" <notifications@github.com> wrote:

@nietras <https://github.com/nietras> Your idea with the workaround has
inspired me :)

You suggested a workaround with deriving non-generic type from generic
type. I realized that we do that in our Engine anyway. And I used this to
overcome ClrMD limitations

After this
<https://ci.appveyor.com/project/dotnetfoundation/benchmarkdotnet/build/0.10.12.420>
build is finished the new package on our CI will contain the fix.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/dotnet/BenchmarkDotNet/issues/631#issuecomment-362911991>,
or mute the thread
